Core: shared clock

algattik opened the pull request:
## What this PR changes/adds

Use a mockable `Clock` to get the time.

Removes calls to `Instant.now()`, `new Date()`, `System.currentTimeMillis()` from production code.

## Why it does that

- Make access to system time consistent in code base.
- Make time mockable for testing.
- Make tests simple and reliable (avoid approximate time comparisons, etc.).

## Further notes

- Moved convenience method implementations from `DefaultServiceExtensionContext` to the interface
- Changed `DataPlaneInstance` contract to return null instead of 0 for empty `lastActive` time.
- `nbf`: in discussion with Ben
- Removed `rollbackState` method in ContractNegotiation and TransferProcess, as it's unused and violates state machine contract.
- Consistently inject services in `ContractServiceExtension`

## Linked Issue(s)

Contributes to #1364 (a decision record will also be provided)

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

paullatzelsperger commented on core/boot/src/main/java/org/eclipse/dataspaceconnector/boot/system/DefaultServiceExtensionContext.java:
```diff
@@ -47,28 +48,14 @@ public DefaultServiceExtensionContext(TypeManager typeManager, Monitor monitor,
         registerService(TypeManager.class, typeManager);
         registerService(Monitor.class, monitor);
         registerService(Telemetry.class, telemetry);
+        registerService(Clock.class, Clock.systemUTC());
```
Unless there is a good reason to the contrary we should provide the clock with a `@Provider(isDefault=true)`-annotated factory method in the `DefaultServicesExtension`.

algattik replied:
Great input, done :)

algattik replied:
The reasons for this are:
- to provide a utility method `ServiceExtensionContext#getClock` so that extensions don't need to inject a clock service. If we move the default clock to `DefaultServicesExtension`, we can't have such a utility method, since we need the dependency resolver to start extensions in the right order.
- that the clock is a common service just like `TypeManager`, `Monitor` and `Telemetry`, and there's no obvious reason to manage it differently

juliapampus commented:
Does this also resolve issue #341? @algattik

jimmarino reviewed (approved):
Great stuff @algattik. I'd like to have a Decision Record since it is important to have this approach documented and everyone aware across the codebase.

Given the conflict issue, that could be done in a subsequent PR, even though it would have been better to commit the DR first.

algattik commented:
> Does this also resolve issue #341? @algattik

@juliapampus no, this is a pure "internal utils" change and does not affect any contracts or behaviours

algattik commented:
> Great stuff @algattik. I'd like to have a Decision Record since it is important to have this approach documented and everyone aware across the codebase.
> 
> Given the conflict issue, that could be done in a subsequent PR, even though it would have been better to commit the DR first.

Great point, I've updated the parent issue to add the need for a DR, and removed the "Closes" comment from this PR. I'll do a DR next.
